Solve the following set of problems using Python <br>
1. Create an MLP model with 16 hidden layers using "mnist_784" dataset from sklearn and improve the result using hyperparameter tuning. Hints <br>
2. Explain your findings and report best hyperparameter values.


## <u>Solution</u>

1. Create an MLP model with 16 hidden layers using "mnist_784" dataset from sklearn and improve the result using hyperparameter tuning.



The dataset "mnist_784" is a large database of handwritten digits which is mainly used for model building for image processing systems. It is a part of a larger dataset NIST, it is under the OpenML community. The data is stored based on the pixel values, it contains 10 different values (0-9)<b>(mnist_784 (2014):Description)</b>.

In [1]:
from sklearn.neural_network import MLPClassifier #NN libraries
from sklearn.datasets import fetch_openml

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score,classification_report

In [2]:
# Load the MNIST dataset
mnist = fetch_openml("mnist_784")
X, y = mnist.data, mnist.target

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [3]:
X

pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
69995     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
69996     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
69997     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
69998     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
69999     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

       pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  \
0          0.0  ...       0.0       0.0       0.0       0.0       0.0   
1          0.0  ...       0.0       0.0       0.0       0.0       0.0   
2          0.0  ...       0.0       0.0       0.0       0.0       0.0   
3          0.0  ...       0.0       0.0       0.0       0.0       0.0   
4          0.0  ...       0.0       0.0       0.0       0.0       0.0   
...        ...  ...       ...       ...       ...       ...       ...   
69995      0.0  ...       0.0       0.0       0.0       0.0       0.0   
69996      0.0  ...       0.0       0.0       0.0       0.0       0.0   
69997      0.0  ...       0.0       0.0       0.0       0.0       0.0   
69998      0.0  ...       0.0       0.0       0.0       0.0       0.0   
69999      0.0  ...       0.0       0.0       0.0       0.0       0.0   

       pixel780  pixel781  pixel782  pixel783  pixel784  
0           0.0       0.0       0.0       0.0       0.0  
1           0.0       0.0       0.0       0.0       0.0  
2           0.0       0.0       0.0       0.0       0.0  
3           0.0       0.0       0.0       0.0       0.0  
4           0.0       0.0       0.0       0.0       0.0  
...         ...       ...       ...       ...       ...  
69995       0.0       0.0       0.0       0.0       0.0  
69996       0.0       0.0       0.0       0.0       0.0  
69997       0.0       0.0       0.0       0.0       0.0  
69998       0.0       0.0       0.0       0.0       0.0  
69999       0.0       0.0       0.0       0.0       0.0  

[70000 rows x 784 columns]

In [4]:
y

0        5
1        0
2        4
3        1
4        9
        ..
69995    2
69996    3
69997    4
69998    5
69999    6
Name: class, Length: 70000, dtype: category
Categories (10, object): ['0', '1', '2', '3', ..., '6', '7', '8', '9']

In [5]:
X.shape

(70000, 784)

It has 70000 datavalues and 784 features.

In [6]:
X.isnull().sum().sum()

0

No null values present

In [7]:
y.value_counts()

1    7877
7    7293
3    7141
2    6990
9    6958
0    6903
6    6876
8    6825
4    6824
5    6313
Name: class, dtype: int64

9 distinct classes, obviously they are numbers.

In [8]:
# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=8)


In [9]:
# We need to scale the data before model building, it is a good practise
scale = StandardScaler()
X_train_scaled = scale.fit_transform(X_train)
X_test_scaled = scale.transform(X_test)

After scaling lets build our model with 16 hidden layers

In [10]:
# Create an MLP model
mlp = MLPClassifier(hidden_layer_sizes=(16,)*16,random_state=8)
mlp.fit(X_train_scaled, y_train)

MLPClassifier(hidden_layer_sizes=(16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16,
                                  16, 16, 16, 16, 16),
              random_state=8)

The hidden_layer_sizes parameter allows us to set the number of layers and number of neurons for each layer. The length of the tuple represent the number of hidden layer <b>(A Beginner’s Guide To Scikit-Learn’s MLPClassifier (2019):Measuring the Accuracy)</b>

In [11]:
# Predict using the initial model
y_predict = mlp.predict(X_test_scaled) #Predicting using the scaled test data

In [16]:
# Calculate accuracy of the initial model
accuracy = accuracy_score(y_test, y_predict)
print("Initial Model Test Accuracy:", accuracy) #The first model accuracy

Initial Model Test Accuracy: 0.9395714285714286


In [17]:
print('Classification report: \n', classification_report(y_test, y_predict))

Classification report: 
               precision    recall  f1-score   support

           0       0.97      0.98      0.97      1430
           1       0.98      0.96      0.97      1557
           2       0.89      0.95      0.92      1366
           3       0.94      0.92      0.93      1406
           4       0.96      0.94      0.95      1398
           5       0.93      0.91      0.92      1273
           6       0.93      0.95      0.94      1326
           7       0.95      0.95      0.95      1487
           8       0.90      0.90      0.90      1323
           9       0.93      0.91      0.92      1434

    accuracy                           0.94     14000
   macro avg       0.94      0.94      0.94     14000
weighted avg       0.94      0.94      0.94     14000



Our model is giving an accuracy of 93.9% which is a good measure. Coming to other performance metrics like precision,recall and f1-score we can see those are also giving good values.

Now lets try to improve the performance of our model by hyperparameter tuning. We are using GridSearchCV to find the optimal parameters.

In [18]:
#New value to try out in the parameter
param_grid = {

    'activation': ['relu', 'tanh'],
    'solver':['adam'],
    'alpha': [0.0001,0.01],
    'max_iter':[100,200]
}

In [19]:
# Perform GridSearchCV for hyperparameter tuning
gS = GridSearchCV(MLPClassifier(), param_grid, cv=3,verbose=3)  #Using the custom parameters to try out and find the error.
gS.fit(X_train_scaled, y_train) #Each combination of parameters will be tried out and its training score is calculated

Fitting 3 folds for each of 8 candidates, totalling 24 fits
[CV 1/3] END activation=relu, alpha=0.0001, max_iter=100, solver=adam;, score=0.970 total time= 1.0min
[CV 2/3] END activation=relu, alpha=0.0001, max_iter=100, solver=adam;, score=0.971 total time= 1.1min
[CV 3/3] END activation=relu, alpha=0.0001, max_iter=100, solver=adam;, score=0.969 total time=  57.9s
[CV 1/3] END activation=relu, alpha=0.0001, max_iter=200, solver=adam;, score=0.970 total time=  58.5s
[CV 2/3] END activation=relu, alpha=0.0001, max_iter=200, solver=adam;, score=0.969 total time= 1.1min
[CV 3/3] END activation=relu, alpha=0.0001, max_iter=200, solver=adam;, score=0.959 total time= 1.0min
[CV 1/3] END activation=relu, alpha=0.01, max_iter=100, solver=adam;, score=0.972 total time= 1.1min
[CV 2/3] END activation=relu, alpha=0.01, max_iter=100, solver=adam;, score=0.972 total time=  52.9s
[CV 3/3] END activation=relu, alpha=0.01, max_iter=100, solver=adam;, score=0.970 total time=  53.4s
[CV 1/3] END activa

GridSearchCV(cv=3, estimator=MLPClassifier(),
             param_grid={'activation': ['relu', 'tanh'],
                         'alpha': [0.0001, 0.01], 'max_iter': [100, 200],
                         'solver': ['adam']},
             verbose=3)

Since we have tuned the parameters lets choose the optimal parameter and use it in our MLPClassifier to predict values to give maximum performance

In [20]:
# Optimal parameter for the model
best_params = gS.best_params_
best_model = gS.best_estimator_

In [21]:
# Model prediction using the optimal parameters we found
y_pred_opt = best_model.predict(X_test_scaled)

In [22]:
# Calculate accuracy of the best model
accuracy_best = accuracy_score(y_test, y_pred_opt)
print("Best Parameters:", best_params)
print("Best Model Test Accuracy:", accuracy_best)

Best Parameters: {'activation': 'relu', 'alpha': 0.01, 'max_iter': 100, 'solver': 'adam'}
Best Model Test Accuracy: 0.9760714285714286


In [23]:
print('Classification report: \n', classification_report(y_test, y_pred_opt))

Classification report: 
               precision    recall  f1-score   support

           0       0.99      0.99      0.99      1430
           1       0.99      0.99      0.99      1557
           2       0.97      0.97      0.97      1366
           3       0.97      0.97      0.97      1406
           4       0.98      0.97      0.98      1398
           5       0.97      0.97      0.97      1273
           6       0.98      0.99      0.98      1326
           7       0.97      0.98      0.98      1487
           8       0.98      0.96      0.97      1323
           9       0.96      0.97      0.97      1434

    accuracy                           0.98     14000
   macro avg       0.98      0.98      0.98     14000
weighted avg       0.98      0.98      0.98     14000



From above we can see that the tuned model is giving almost 98% accuracy which is far better than the first model with default parameters.

2. Explain your findings and report best hyperparameter values.

In [24]:
#The best hyperparameter values are:
print("Best Parameters:", best_params)

Best Parameters: {'activation': 'relu', 'alpha': 0.01, 'max_iter': 100, 'solver': 'adam'}


With the default MLP Classifier with 16 hidden layers and neurons we only got 93.9% accuracy but with hyperparameter tuning using GridSearchCV we were able to produce an accuracy of 97.6%. From the classification report we can see that the individual scores of each class also improved as a result of tuning.<br>

The optimal parameters we received consisted of using 'relu'(Rectified Linear Activation) as activation function this is also the default value. <br>

The Regularization Strength or alpha got the value as 0.01, for which the default value is 0.0001. The value increased which means a small regularization was effective for our result.<br>

The model achieved its convergence within the maximum iteration of 100, the default value for this is 200 which was extra than needed.<br>

We only used 'adam' optimizer which is also the default value because it is an adaptive optimization algorithm that changes the learning rate accordingly during the training of the model.<br>
<br>

Hyperparameter tuning is crucial step in building accurate machine learning models. It gives more impact than we consider.

## <u>Reference List</u>

Analytics India Mag (20 June 2019)A Beginner’s Guide To Scikit-Learn’s MLPClassifier, accessed 16 August 2023.<br>
https://analyticsindiamag.com/a-beginners-guide-to-scikit-learns-mlpclassifier/<br><br>

OpenML (29 September 2014)mnist_784, accessed 16 Augest 2023.<br> https://www.openml.org/search?type=data&sort=runs&id=554&status=active <br><br>